runtime/compiler/z/codegen/J9TreeEvaluator.cpp

```diff
@@ -3396,6 +3425,664 @@ J9::Z::TreeEvaluator::ArrayCopyBNDCHKEvaluator(TR::Node * node, TR::CodeGenerato
+   bool isStatic = symRef->getSymbol()->getKind() == TR::Symbol::IsStatic;
+
+   TR_RuntimeHelper helperIndex = isWrite? (isStatic ?  TR_jitResolveStaticFieldSetterDirect: TR_jitResolveFieldSetterDirect):
+                                          (isStatic ?  TR_jitResolveStaticFieldDirect: TR_jitResolveFieldDirect);
+
+   TR::S390HelperLinkage *helperLink = static_cast<TR::S390HelperLinkage*>(cg->getLinkage(runtimeHelperLinkage(helperIndex)));
+
+   intptr_t offsetInDataBlock = isStatic ? offsetof(J9JITWatchedStaticFieldData, fieldAddress): offsetof(J9JITWatchedInstanceFieldData, offset);
+
+   TR::LabelSymbol* startLabel = generateLabelSymbol(cg);
+   TR::LabelSymbol* endLabel = generateLabelSymbol(cg);
+   TR::LabelSymbol *unresolvedLabel = generateLabelSymbol(cg);
+   startLabel->setStartInternalControlFlow();
+   endLabel->setEndInternalControlFlow();
+
+   // need 2 registers for arguments, 2 for entry point and return address register, and 1 for datablock
+   uint8_t numOfConditions = 5;
+
+   if (isStatic) // needs fieldClassReg
+      numOfConditions++;
+
+   TR::RegisterDependencyConditions  *deps =  generateRegisterDependencyConditions(0, numOfConditions, cg);
+   TR::Register *resultReg = NULL;
+   TR::Register *tempReg = cg->allocateRegister();
+   TR::Register *dataBlockReg = cg->allocateRegister();
+
+   deps->addPostCondition(dataBlockReg, TR::RealRegister::AssignAny);
+
+   generateRILInstruction(cg, TR::InstOpCode::LARL, node, dataBlockReg, dataSnippet);
+   generateRILInstruction(cg, TR::InstOpCode::LGFI, node, tempReg, static_cast<int32_t>(-1));
+
+   generateRXInstruction(cg, is64Bit ? TR::InstOpCode::CG : TR::InstOpCode::C, node, tempReg, generateS390MemoryReference(dataBlockReg, offsetInDataBlock, cg));
+
+   generateS390BranchInstruction(cg, TR::InstOpCode::BRC, TR::InstOpCode::COND_BNE, node, endLabel); // branch to end if not set
+   generateS390LabelInstruction(cg, TR::InstOpCode::LABEL, node, startLabel);
+     
+   if (isStatic)
+      {
+      // Fills in J9JITWatchedStaticFieldData.fieldClass
+      TR::Register *fieldClassReg = cg->evaluate(isWrite ? getDirectWrtbarJ9ClassNode(node): node->getFirstChild());
+      generateRXInstruction(cg, TR::InstOpCode::getStoreOpCode(), node, fieldClassReg, generateS390MemoryReference(dataBlockReg, (int32_t)(offsetof(J9JITWatchedStaticFieldData, fieldClass)), cg));
+      deps->addPostCondition(fieldClassReg, TR::RealRegister::AssignAny);
+      }
+   TR::ResolvedMethodSymbol *methodSymbol = node->getByteCodeInfo().getCallerIndex() == -1 ? cg->comp()->getMethodSymbol(): cg->comp()->getInlinedResolvedMethodSymbol(node->getByteCodeInfo().getCallerIndex());
+        
+   // setup linkage 
+   TR::Register *cpAddressReg = cg->allocateRegister();
+   TR::Register *cpIndexReg = cg->allocateRegister();
+
+   generateRegLitRefInstruction(cg, TR::InstOpCode::LG, node, cpAddressReg, (uintptrj_t) methodSymbol->getResolvedMethod()->constantPool(), TR_ConstantPool, deps, 0, 0);
```

Comment:
This doesn't look correct for 32-bit. Are constant pool addresses going to be 32-bit values in that case?

Reply:
This was fixed here: https://github.com/eclipse/openj9/pull/4743/commits/1d9fce5a23abacb0a12dbff7a6f919251f4d0fbe